### ISSUE TITLE
Error reporting for incorrect return types should be more consistent

### ISSUE THREAD

**Issue by kevinbarabash** [microsoft/TypeScript#42161]
# Suggestion

## 🔍 Search Terms

`return type label:"Domain: Error Messages"`

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Error reporting for mismatches between a functions return type and what is being returned by a particular `return` statement should be consistently reported a the location of the offending `return`.

## 📃 Motivating Example

```typescript
type Foo = (input: string) => number;

const foo: Foo = (input) => {
  switch (input) {
    case "1":
      return 1;
    case "2":
      return "2";
    default:
      return Infinity;
  }
}
```

The following error is reported on the declaration of `foo`.
```
const foo: Foo
Type '(input: string) => number | "2"' is not assignable to type 'Foo'.
  Type 'number | "2"' is not assignable to type 'number'.
    Type 'string' is not assignable to type 'number'.(2322)
```

```typescript
const bar = (input: string): number => {
  switch (input) {
    case "1":
      return 1;
    case "2":
      return "2";
    default:
      return Infinity;
  }
}
```

The following error is reported on `return "2"`:
```
Type 'string' is not assignable to type 'number'.(2322)
```

## 💻 Use Cases

In the second example it's much easier to see where the problem is.  In larger functions for more `returns it can be even hard to pin down which one is causing the issue.  These two examples are very similar and I would expect the same error to be reported for each.

**Comment by andrewbranch** [comment]
I don’t think we would ever get rid of the error where it’s placed in the first example. The fundamental assignability check there is the whole function to `Foo`, whereas in the latter, the assignability check is the return type to `number`. I would argue that being able to follow the logic of these comparisons is an acquired skill, but is actually useful and important to users as they try to work out increasingly complicated errors. However, I totally agree that once the error elaboration gets down to the `"2"` comparison, it can be hard to find what it’s talking about if the function is any bigger. It seems like there might be a good opportunity for a related error span here. /cc @DanielRosenwasser, as error reporting is his favorite topic.

**Comment by DanielRosenwasser** [comment]
To me, reporting the error on the `"2"` is probably the most ideal thing - at least in the above 2 scenarios. The problem comes down to what Andrew said - we're not considering every piece of the program while doing that comparison, and in the first example, the function looks at its return values for its return type. It's very similar to the following snippet:

```ts
declare function foo(f: () => string): void;

foo(() => {
    return 100;
})
```

**Comment by DanielRosenwasser** [comment]
So I'm not totally sure what the right thing to do is; you could imagine that in the presence of every function expression/arrow function being assigned to something with a single call signature, we do a non-reporting check, and if that fails, gather up return expressions from the source and check against the return type of the target each time.

**Comment by RyanCavanaugh** [comment]
The implementation is likely to be tricky but I think the desirable behavior is pretty clear here - shall we tag Experience Enhancement?

**Comment by DanielRosenwasser** [comment]
The difference between the two is pretty hazy to me 😅 but I don't mind `Experience Enhancement`.

**Comment by kevinbarabash** [comment]
Yeah, when to report what kind of error can be pretty contextual.  Explicitly including the return value even when using the `Foo` type gives me the behavior I want in this case.  It's a little extra typing so not a big deal.
